### PR TITLE
New exercise: generic sorting.

### PR DESCRIPTION
**Pull request by fpottier**
Generic sorting and discrimination, based on Henglein's paper.
These linear-time algorithms work for every preorder in a family that is described by a GADT, `'k order`.

### CODE REVIEW

File: exercises/generic_sorting/descr.md [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+there is a linear-time algorithm that sorts values of type `'k2` according to
+a preorder `o2`, then it is not very difficult to construct a linear-time
+algorithm that sorts **pairs** of type `'k1 * 'k2` according to the
+lexicographic ordering `o1 * o2`.
+
+These remarks suggest that there is **a family of preorders** for which
+one can construct a linear-time sorting algorithm.
+
+In this exercise, we exhibit such a family of preorders. It takes the form of
+an algebraic data type `'k order`. A value `o` of type `'k order` is a
```
Comment (yurug) [inline exercises/generic_sorting/descr.md]: Why didn't you name this type `preorder`?

File: exercises/generic_sorting/descr.md [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+```
+  val sort: 'k order -> ('k * 'v) list -> 'v list
+```
+
+The function `sort o` accepts a list of key-value pairs,
+sorts it according to the preorder `o`,
+which is a preorder on keys,
+and returns just a list of values.
+The keys are dropped.
```
Comment (yurug) [inline exercises/generic_sorting/descr.md]: Maybe a word about this design choice? Why don't we simply type `sort` as follows:
```
sort: 'a order -> 'a list -> 'a list
```
?

File: exercises/generic_sorting/descr.md (hunk 1 above)
Comment (fpottier) [reply exercises/generic_sorting/descr.md]: It's a common abuse of language, I think, to say "order" where a preorder is allowed. For instance, OCaml's `Set.Make` is parameterized with an `OrderedType` -- a type equipped with a preorder.

If you would like, I can make the change. It's just a bit painful as there are many occurrences.

File: exercises/generic_sorting/descr.md (hunk 2 above)
Comment (fpottier) [reply exercises/generic_sorting/descr.md]: I have added an explanation.

File: exercises/generic_sorting/descr.md (hunk 1 above)
Comment (yurug) [reply exercises/generic_sorting/descr.md]: Keep it that way.

File: exercises/generic_sorting/descr.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+(This somewhat unusual presentation, where keys and values are distinguished,
+helps define `sort` recursively. A more standard presentation, where keys and
+values are identified, can easily be obtained as a special case; see Question 4.)
+
+The time complexity of `sort o kvs` is `O(n)`, where `n` is the length of the
+input list `kvs`, and where the constant factor is allowed to depend on the
+preorder `o`.
```
Comment (yurug) [inline exercises/generic_sorting/descr.md]: What does that mean to have a constant factor depending on a preorder? Can this dependency be a way to cheat about the complexity?

File: exercises/generic_sorting/descr.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+*Note.* The automatic grading code verifies whether the output list is
+correct, but does not verify the time complexity of your algorithm. It
+is up to you to convince yourself that the complexity bound is met.
+
+*Note.* In a real-world implementation,
+one would use memoization so as to ensure
+that two successive calls to `pigeonhole_sort bound`
+use the same internal array,
+instead of allocating a new array every time.
```
Comment (yurug) [inline exercises/generic_sorting/descr.md]: That's not clear to me as this would prevent concurrent executions of this function.

Comment (fpottier) [reply exercises/generic_sorting/descr.md]: OK, reformulated in a less affirmative way.

File: exercises/generic_sorting/descr.md (hunk 3 above)
Comment (fpottier) [reply exercises/generic_sorting/descr.md]: I have tried to clarify. Basically the constant factor is the syntactic size of the preorder `o`. However, `OMap` nodes complicate things, as they could carry a function whose cost is arbitrary. So I add the assumption that every such function has constant cost.

**Pull request merged**